File: exercises/binary_tree.php
<?php

class Node {
    function __construct(
        public mixed $value,
        public ?Node $left = null,
        public ?Node $right = null,
    ) {

    }
}

function tree_to_value_array_depth_first(?Node $root): array {
    $values = [];
    if ($root !== null) {
        $values[] = $root->value;

        if ($root->left !== null) {
            $values = array_merge($values, tree_to_value_array_depth_first($root->left));
        }
        if ($root->right !== null) {
            $values = array_merge($values, tree_to_value_array_depth_first($root->right));
        }
    }
    return $values;
}

function tree_to_value_array_breadth_first(?Node $root): array {
    $values = [];
    $queue = new SplQueue();

    if ($root !== null) {
        $queue->enqueue($root);
    }

    while (!$queue->isEmpty()) {
        /** @var Node $currNode */
        $currNode = $queue->dequeue();
        $values[] = $currNode->value;

        if ($currNode->left !== null) {
            $queue->enqueue($currNode->left);
        }

        if ($currNode->right !== null) {
            $queue->enqueue($currNode->right);
        }
    }

    return $values;}

File: exercises/binary_tree.py
from __future__ import annotations

# class Node:
#     def __init__(
#         self,
#         value,
#         left: Node|None = None,
#         right: Node|None = None
#     ):
#         self.value = value
#         self.left = left
#         self.right = right

from dataclasses import dataclass
from typing import Generic, TypeVar, List

T = TypeVar("T")

@dataclass(slots=True)
class Node(Generic[T]):
    value: T
    left: Node[T]|None = None
    right: Node[T]|None = None

def tree_to_value_array_depth_first(root: Node[T]|None) -> List[T]:
    values: List[T] = []

    if root:
        values.append(root.value)
        values.extend(tree_to_value_array_depth_first(root.left))
        values.extend(tree_to_value_array_depth_first(root.right))

    return values

def tree_to_value_array_breadth_first(self, root: Node[T]|None)->List[T]:
    values: List[T] = []
    from collections import deque
    queue: deque = deque()
    if root:
        queue.append(root)
    
    while queue:
        currNode: Node = queue.popleft()
        values.append(currNode.value)

        if currNode.left is not None:
            queue.append(currNode.left)

        if currNode.right is not None:
            queue.append(currNode.right)                           
    return values